pkg/controllers/membercluster/v1beta1/membercluster_controller.go

```diff
@@ -115,42 +117,61 @@ func (r *Reconciler) Reconcile(ctx context.Context, req runtime.Request) (runtim
 
 // handleDelete handles the delete event of the member cluster, makes sure the agent has finished leaving the fleet first and
 // then garbage collects all the resources in the cluster namespace.
-func (r *Reconciler) handleDelete(ctx context.Context, mc *clusterv1beta1.MemberCluster) (runtime.Result, error) {
+func (r *Reconciler) handleDelete(ctx context.Context, mc *clusterv1beta1.MemberCluster) error {
 	mcObjRef := klog.KObj(mc)
 	if !controllerutil.ContainsFinalizer(mc, placementv1beta1.MemberClusterFinalizer) {
 		klog.V(2).InfoS("No need to do anything for the deleting member cluster without a finalizer", "memberCluster", mcObjRef)
-		return runtime.Result{}, nil
+		return nil
 	}
-	currentImc, err := r.getInternalMemberCluster(ctx, mc.GetName())
-	if err != nil {
-		return runtime.Result{}, err
+	// check if the namespace still exist
+	var currentNS corev1.Namespace
+	namespaceName := fmt.Sprintf(utils.NamespaceNameFormat, mc.Name)
+	if err := r.Client.Get(ctx, types.NamespacedName{Name: namespaceName}, &currentNS); err != nil {
+		if !apierrors.IsNotFound(err) {
+			klog.ErrorS(err, "failed to get the member cluster namespace", "memberCluster", mcObjRef)
+			return controller.NewAPIServerError(true, err)
+		}
+		klog.V(2).InfoS("the member cluster namespace is not found, remove the finalizer", "memberCluster", mcObjRef)
+		controllerutil.RemoveFinalizer(mc, placementv1beta1.MemberClusterFinalizer)
+		return controller.NewUpdateIgnoreConflictError(r.Update(ctx, mc, &client.UpdateOptions{}))
+	}
+	currentImc := &clusterv1beta1.InternalMemberCluster{}
+	imcNamespacedName := types.NamespacedName{Namespace: namespaceName, Name: mc.Name}
+	if err := r.Client.Get(ctx, imcNamespacedName, currentImc); err != nil {
+		if !apierrors.IsNotFound(err) {
+			klog.ErrorS(err, "failed to get internal member cluster", "internalMemberCluster", imcNamespacedName)
+			return controller.NewAPIServerError(true, err)
+		}
+		// we don't need to wait for the agent to leave if the internal member cluster is not found
+		klog.ErrorS(controller.NewUnexpectedBehaviorError(fmt.Errorf("internal member cluster %s not found", mcObjRef)), "InternalMemberCluster not found, start garbage collecting", "memberCluster", mcObjRef)
```

Comment:
now it becomes expected error? cause mc will be there until we delete all the namespace scoped resources and imc may be gone.